src/libvirt-xml-update.js

```diff
@@ -296,3 +296,56 @@ export function updateMaxMemory(domXml, maxMemory) {
 
     return s.serializeToString(doc);
 }
+
+function xmlGetXPath(doc, xpath) {
+    const elements = [];
+    let node = null;
+    const iterator = doc.evaluate(xpath, doc);
+    while ((node = iterator.iterateNext()))
+        elements.push(node);
+    return elements;
+}
+
+export function replaceSpice(domXml) {
+    // this implements https://access.redhat.com/solutions/6955095
+    logDebug("replaceSpice original XML:", domXml);
+
+    const doc = getDoc(domXml);
+
+    xmlGetXPath(doc, "/domain/devices/audio[@type='spice']").forEach(n => n.remove());
+    xmlGetXPath(doc, "//domain/devices/redirdev[@type='spicevmc']").forEach(n => n.remove());
+    xmlGetXPath(doc, "//domain/devices/channel[@type='spicevmc']").forEach(n => n.remove());
+
+    xmlGetXPath(doc, "//domain/devices/video").forEach(video => {
+        const model = getSingleOptionalElem(video, "model");
+        if (model?.getAttribute("type") === "qxl") {
+            // remove all attributes except for primary; don't change list while iterating
+            const removeAttrs = [];
+            for (let i = 0; i < model.attributes.length; i++) {
+                const attr = model.attributes.item(i);
+                if (attr.name !== "primary")
+                    removeAttrs.push(attr.name);
+            }
+            removeAttrs.forEach(attr => model.removeAttribute(attr));
+            model.setAttribute("type", "vga");
+        }
+    });
+
+    const spice_graphics = xmlGetXPath(doc, "//domain/devices/graphics[@type='spice']");
+    // do we already have a VNC graphics device?
+    if (xmlGetXPath(doc, "//domain/devices/graphics[@type='vnc']").length > 0) {
+        // then simply remove the spice graphics device
+        spice_graphics.forEach(graphics => graphics.remove());
+    } else {
+        // otherwise replace spice graphics with VNC
```

Comment:
Say there are multiple spice entries. Then, if there is a single vnc entry, all spice entries are removed, leaving the single vnc entry. But if there is no vnc entry, each spice entry is converted into a vnc entry, leaving multiple vnc entries. Also, spice specific elements might still be in those VNC entries. That seems hairy.

What about always removing all spice entries, and then adding a generic vnc entry if there isn't one already?

Reply:
There can only be one `<graphics>` of each type, otherwise libvirt will complain. So in *practice* this should be fine, but conceptually you are right. I'm just a bit wary of completely deleting the spice device, as they often have things like virtual PCI slot, or the IP to listen on, and I'm e.g. afraid of replacing a localhost or "my internal network" IP with a hardcoded one. Essentially I'm implementing the [official documentation](https://access.redhat.com/solutions/6955095). This would really be something where I'd appreciate the libvirt team's input. I asked in the Jira ticket a few days ago.

Reply:
> Also, spice specific elements might still be in those VNC entries

If there are entries which we don't recognize, I specifically *want* the auto-migration to fail, rather than stomping over the admin's customizations. I'm ok with converting known cases, but if we encounter something which we don't understand (i.e. when `virt-xml-validate` says :-1: ), we revert and say "sorry, go do it yourself".

Reply:
@martinpitt yes I think `s/type='spice'/type='vnc'/` is the sensible way to handle it for `<graphics>`. For common use cases it should do the right thing.

Libvirt doesn't heavily validate `<graphics>` XML like it does for some other devices; most of the config bits that work for spice and not VNC will be thrown away at XML parse time. `<listen>` config works mostly the same for both types though

Reply:
@crobinso Thanks. Just to make sure: is it ever a thing to have more than one spice graphics device? My gut feeling is that doing an 1:1 conversion of spice to vnc is conceptually cleaner than removing them all and adding back a standard one, as that silently removes customizations.

Reply:
> I'm ok with converting known cases, but if we encounter something which we don't understand (i.e. when virt-xml-validate says 👎 ), we revert and say "sorry, go do it yourself".

How are we handling error cases? Will the VM still boot? (It would be a warning then, which should probably be dismissable if it's not a problem. If it is a problem, then we need to show it until it's not a problem, and provide some guidance on how to fix it, if we cannot already easily fix it from within Cockpit.)

Reply:
If the conversion fails, the VM remains unchanged -- i.e. it will still boot. You'd just get an error message in the dialog:

![image](https://github.com/cockpit-project/cockpit-machines/assets/200109/0ca6b91c-8531-4fd9-99eb-1fe59548a59c)

At this point there isn't an automatic solution any more. But for the record, we don't currently know of a current case for this -- to test this we need to explicitly fake/break `virt-xml-validate`.

Reply:
After someone fixes the problem, does the error message automatically go away?

Reply:
No, it doesn't -- it's a dialog, at this point you can only cancel it. I.e. at this point the user only has the options to convert the XML manually (following https://access.redhat.com/solutions/6955095 but it *will* be incomplete as we implemented all the steps there), or reconsider and maybe move the VM to a host which does support spice. At this point we can really only throw our hands in the air, I'm afraid. (But again: This should be an exceptionally rare scenario)

Reply:
Oh, sorry; for whatever reason, I thought it was at the top of the list after the modal was finished. OK. That works! Thanks!

I guess ordinarily, it'll be near-instant unless it hits an error like this?

Reply:
Yes, both failure and success are near-instant. There is just some XML mangling going on, no variable-length or expensive operation.

Reply:
> @crobinso Thanks. Just to make sure: is it ever a thing to have more than one spice graphics device? My gut feeling is that doing an 1:1 conversion of spice to vnc is conceptually cleaner than removing them all and adding back a standard one, as that silently removes customizations.

Libvirt rejects multiple `<graphics>` with the same `type` value so I don't think you have to worry about that case. In theory someone could have a spice and vnc device already configured, in which case converting spice to another vnc device will fail validation since 2x vnc isn't allowed, but IMO that's obscure enough to leave it for the 'we tried, fix it yourself pls' case

Reply:
We regularly encounter that case, and this PR even covers it with an integration test. That's why we have to do the if/else here, and `virt-xml` will have to do the same.

Thanks for reviewing and confirming! So I suppose this is good enough.

Reply:
Thanks, I get it now, too. Just one thing: Do we need to explicitly look for "vnc" as the alternative? Or should we be ok to remove all spice entries as long as there is _any_ other graphics entry? Maybe vnc is going to cause problems? *shrug* I guess that's very unlikely.

Reply:
That's a question for @crobinso or @germanovm -- I understand the [request](https://issues.redhat.com/browse/RHEL-17434) was rather specific about converting spice to VNC consoles. I also agree that it makes sense -- e.g. having an additional local `sdl` or `gtk` console won't help you much as the point is that we want to keep a *remote* console.